### PR TITLE
Turn on smarter FakeDockerClient in HollowNode.

### PR DESCRIPTION
**Pull request by gmarek**
cc @wojtek-t @yujuhong

### CODE REVIEW

File: pkg/kubelet/dockertools/fake_docker_client.go [hunk 1]
```diff
@@ -175,27 +176,36 @@ func (f *FakeDockerClient) InspectImage(name string) (*docker.Image, error) {
 	return f.Image, err
 }
 
+func normalSleep(mean, stddev, cutoffMillis int) {
```
Comment (yujuhong) [inline pkg/kubelet/dockertools/fake_docker_client.go]: Could you add an sentence here explaining how this function should be used? The unit of the variables is also unclear.

**Comment by yujuhong** [comment]
Since this is only useful for benchmarking, could we add a knob to turn it off (i.e., no sleep) by default? The fake docker client is used in unit tests where the extra delay is not really appreciated.

Also, I'd appreciate if you can incorporate https://github.com/yujuhong/kubernetes/commit/ba83f5dbf086dc9f86ccd16e95bb0face00fcbda when you have time :)

File: pkg/kubelet/dockertools/fake_docker_client.go (hunk 1 above)
Comment (gmarek) [inline pkg/kubelet/dockertools/fake_docker_client.go]: Done.

**Comment by gmarek** [comment]
I added a knob, but I don't have access to workstation so incorporating anything would take a lot of time. And I really need to get this in, to have few jenkins runs in the morning when I come to work.

File: pkg/kubelet/dockertools/fake_docker_client.go [hunk 2]
```diff
@@ -175,27 +177,40 @@ func (f *FakeDockerClient) InspectImage(name string) (*docker.Image, error) {
...
+	// This is not a very good fake. We'll just add this container's name to the list.
+	// Docker likes to add a '/', so copy that behavior.
+	name := "/" + c.Name
+	f.ContainerList = append(f.ContainerList, docker.APIContainers{ID: name, Names: []string{name}, Image: c.Config.Image})
+	container := docker.Container{ID: name, Name: name, Config: c.Config}
+	if f.ContainerMap != nil {
+		containerCopy := container
+		f.ContainerMap[name] = &containerCopy
+	}
+	if f.EnableSleep {
```
Comment (yujuhong) [inline pkg/kubelet/dockertools/fake_docker_client.go]: nit: making `normalSleep` a method and checks the `f.EnableSleep` there; this avoids checking before use every time.

**Comment by yujuhong** [comment]
> I added a knob, but I don't have access to workstation so incorporating anything would take a lot of time. And I really need to get this in, to have few jenkins runs in the morning when I come to work.

Thanks for the fix. LGTM with one nit.

BTW, did you benchmark docker to get the those sleep time numbers? I'd like to know how you categorize the "important" functions. Thanks.

**Comment by yujuhong** [comment]
@gmarek, since it's just a nit, I'll trust you to fix it later. 

Is the kubemark work slotted for v1.1? I'd like to know so we could coordinate relevant code reviews better. Thanks.

**Comment by gmarek** [comment]
I chose important functions more or less arbitrarily, and I added the same sleep everywhere in such amount that the number of Pods started in 10 seconds in the density '30 run is pretty much the same as in normal cluster. It was good enough as a first approximation, and I need more data to tweak it further.

**Comment by yujuhong** [comment]
> I chose important functions more or less arbitrarily, and I added the same sleep everywhere in such amount that the number of Pods started in 10 seconds in the density '30 run is pretty much the same as in normal cluster. It was good enough as a first approximation, and I need more data to tweak it further.

We do record docker requests metrics using prometheus in kubelet. Did you try using those or have you found them not representative? Docker performance would also degrade quite a bit under heavy operations such as pod batch startup. Have you considered adjusting the delays based on the qps?

**Comment by gmarek** [comment]
I'm not at that level of sophistication yet:)

**Comment by gmarek** [comment]
@k8s-bot test this please

**Pull request merged**